R/int_pctl.R

```diff
@@ -0,0 +1,224 @@
+
+  `%op%` <- get_int_p_operator(allow_par)
+  rs$.metrics <-
+    foreach::foreach(
+      i = 1:nrow(rs),
+      .errorhandling = "pass",
+      .packages = c("tune", "rsample")
+    )  %op% {
+      comp_metrics(rs$splits[[i]], y_nm, metrics, event_level)
+    }
+
+  if (any(grepl("survival", .get_tune_metric_names(x)))) {
+    # compute by evaluation time
+    res <- int_pctl_dyn_surv(rs, allow_par)
+  } else {
+    res <- rsample::int_pctl(rs, .metrics)
+  }
+  res %>%
+    dplyr::mutate(.estimator = "bootstrap") %>%
+    dplyr::rename(.metric = term) %>%
+    dplyr::relocate(.estimator, .after = .metric) %>%
+    dplyr::select(-.alpha, -.method) %>%
+    cbind(config)
+}
+
+# We have to do the analysis separately for each evaluation time.
+int_pctl_dyn_surv <- function(x, allow_par) {
+  `%op%` <- get_int_p_operator(allow_par)
+  times <- unique(x$.metrics[[1]]$.eval_time)
+  res <-
+    foreach::foreach(
+      i = seq_along(times),
+      .errorhandling = "pass",
+      .packages = c("purrr", "rsample", "dplyr")
+    )  %op% {
+      by_eval_time(times[i], x)
+    }
+  dplyr::bind_rows(res)
+}
+
+by_eval_time <- function(time, x) {
+  times <- dplyr::tibble(.eval_time = time)
+  x$.metrics <- purrr::map(x$.metrics, ~ dplyr::inner_join(.x, times, by = ".eval_time"))
+  rsample::int_pctl(x, .metrics)
+}
+
+
+# ------------------------------------------------------------------------------
+
+get_configs <- function(x, parameters = NULL, as_list = TRUE) {
+  param <- .get_tune_parameter_names(x)
+  config_cols <- c(".config", ".iter", param)
+  config_keys <-
+    collect_metrics(x, summarize = FALSE) %>%
+    dplyr::distinct(dplyr::pick(dplyr::any_of(config_cols)))
+  if (!is.null(parameters)) {
+    merge_cols <- intersect(names(config_keys), names(parameters))
+    config_keys <- dplyr::inner_join(config_keys, parameters, by = merge_cols)
+  }
+  if (as_list) {
+    config_keys <- vctrs::vec_chop(config_keys, as.list(1:nrow(config_keys)))
+  }
+  config_keys
+}
+
+# Compute metrics for a specific configuration
+comp_metrics <- function(split, y, metrics, event_level) {
+  dat <- rsample::analysis(split)
+  info <- metrics_info(metrics)
+
+  res <-
+    .estimate_metrics(
+      dat,
+      metric = metrics,
+      param_names = NULL,
+      outcome_name = y,
+      event_level = event_level,
+      metrics_info = info,
+      eval_time = NA  # TODO I don't think that this is used in the function
+    )
+
+  res %>%
+    dplyr::rename(term = .metric, estimate = .estimate) %>%
+    dplyr::select(-.estimator)
+}
+
+# ------------------------------------------------------------------------------
+
+
+filter_eval_times <- function(x, eval_time = NULL) {
+  if (is.null(eval_time)) {
+    return(x$.predictions)
+  }
+  purrr::map(x$.predictions, thin_time, times = eval_time)
+}
+
+thin_time <- function(x, times) {
+  x$.pred <- purrr::map(x$.pred, subset_time, times = times)
+  x
+}
+
+subset_time <- function(x, times) {
```

Comment:
This could live inside of `thin_time()` since it's so short. If we want to re-use `subset_time()` elsewhere we can move it outside of `thin_time()` again later.